### PR TITLE
client: fix vulnerabilities found in  `npm audit`

### PR DESCRIPTION
**Pull request by zonggen**
Fixed previous vulnerabilities returned by `npm audit`, after the fix:

```
$ npm audit
                                                                                
                       === npm audit security report ===                        
                                                                                
found 0 vulnerabilities
 in 25779 scanned packages
```

Closes: https://issues.redhat.com/browse/GRPA-1795
Signed-off-by: Allen Bai <abai@redhat.com>

### CODE REVIEW

**Comment by zonggen** [comment]
Fixed by running `npm audit --fix`, manually `npm install --save-dev` and `npm install core-js@3`.

Built, pulled and ran a container (https://quay.io/repository/zonggen/fcct-online/build/861b84ab-233f-49b2-8aa6-bdf91851b4bd), no issues occurred except +0.6MB on size of container.

cc @ashcrow

File: client/package-lock.json [hunk 1]
```diff
@@ -1019,395 +2548,689 @@
...
-          "resolved": "https://registry.npmjs.org/acorn/-/acorn-6.2.1.tgz",
-          "integrity": "sha512-JD0xT5FCRDNyjDda3Lrg/IxFscp9q4tiYtxE1/nOzlKCk7hIRuYjhq1kCNkbPjMRMZuFq20HNQn1I9k8Oj0E+Q==",
-          "dev": true
+          "requires": {
+            "cacache": "^13.0.1",
+            "find-cache-dir": "^3.2.0",
+            "jest-worker": "^25.1.0",
+            "p-limit": "^2.2.2",
+            "schema-utils": "^2.6.4",
+            "serialize-javascript": "^2.1.2",
```
Comment (ashcrow) [inline client/package-lock.json:3149]: :+1:

File: client/package-lock.json [hunk 2]
```diff
@@ -3151,53 +5103,25 @@
...
-        "loader-utils": "^1.1.0",
+        "is-glob": "^4.0.1",
+        "loader-utils": "^1.2.3",
         "minimatch": "^3.0.4",
-        "p-limit": "^1.0.0",
-        "serialize-javascript": "^1.4.0"
+        "normalize-path": "^3.0.0",
+        "p-limit": "^2.2.1",
+        "schema-utils": "^1.0.0",
+        "serialize-javascript": "^2.1.2",
```
Comment (ashcrow) [inline client/package-lock.json:5121]: :+1:

File: client/package-lock.json [hunk 3]
```diff
@@ -9780,9 +11990,9 @@
       }
     },
     "serialize-javascript": {
-      "version": "1.7.0",
-      "resolved": "https://registry.npmjs.org/serialize-javascript/-/serialize-javascript-1.7.0.tgz",
-      "integrity": "sha512-ke8UG8ulpFOxO8f8gRYabHQe/ZntKlcig2Mp+8+URDP1D8vJZ0KUt7LYo07q25Z/+JVSgpr/cui9PIp5H6/+nA==",
+      "version": "2.1.2",
```
Comment (ashcrow) [inline client/package-lock.json:11993]: :+1:

File: client/package-lock.json [hunk 4]
```diff
@@ -10624,39 +12777,37 @@
...
       "requires": {
-        "cacache": "^11.3.2",
-        "find-cache-dir": "^2.0.0",
+        "cacache": "^12.0.2",
+        "find-cache-dir": "^2.1.0",
         "is-wsl": "^1.1.0",
-        "loader-utils": "^1.2.3",
         "schema-utils": "^1.0.0",
-        "serialize-javascript": "^1.7.0",
+        "serialize-javascript": "^2.1.2",
```
Comment (ashcrow) [inline client/package-lock.json:12789]: :+1:

File: client/package-lock.json [hunk 5]
```diff
@@ -4203,9 +6213,9 @@
       },
       "dependencies": {
         "acorn": {
-          "version": "6.2.1",
-          "resolved": "https://registry.npmjs.org/acorn/-/acorn-6.2.1.tgz",
-          "integrity": "sha512-JD0xT5FCRDNyjDda3Lrg/IxFscp9q4tiYtxE1/nOzlKCk7hIRuYjhq1kCNkbPjMRMZuFq20HNQn1I9k8Oj0E+Q==",
+          "version": "6.4.1",
```
Comment (ashcrow) [inline client/package-lock.json:6216]: :+1:

File: client/package-lock.json [hunk 6]
```diff
@@ -4605,9 +6615,9 @@
...
```
Comment (ashcrow) [inline client/package-lock.json:6618]: :+1:

File: client/package-lock.json [hunk 7]
```diff
@@ -11391,124 +13557,144 @@
...
+      "version": "4.42.0",
+      "resolved": "https://registry.npmjs.org/webpack/-/webpack-4.42.0.tgz",
+      "integrity": "sha512-EzJRHvwQyBiYrYqhyjW9AqM90dE4+s1/XtCfn7uWg6cS72zH+2VPFAlsnW0+W0cDi0XRjNKUMoJtpSi50+Ph6w==",
+      "dev": true,
+      "requires": {
+        "@webassemblyjs/ast": "1.8.5",
+        "@webassemblyjs/helper-module-context": "1.8.5",
+        "@webassemblyjs/wasm-edit": "1.8.5",
+        "@webassemblyjs/wasm-parser": "1.8.5",
+        "acorn": "^6.2.1",
```
Comment (ashcrow) [inline client/package-lock.json:13569]: :+1:

File: client/package-lock.json [hunk 8]
```diff
@@ -11391,124 +13557,144 @@
...
-      "resolved": "https://registry.npmjs.org/webpack-bundle-analyzer/-/webpack-bundle-analyzer-3.3.2.tgz",
-      "integrity": "sha512-7qvJLPKB4rRWZGjVp5U1KEjwutbDHSKboAl0IfafnrdXMrgC0tOtZbQD6Rw0u4cmpgRN4O02Fc0t8eAT+FgGzA==",
+      "version": "3.6.1",
+      "resolved": "https://registry.npmjs.org/webpack-bundle-analyzer/-/webpack-bundle-analyzer-3.6.1.tgz",
+      "integrity": "sha512-Nfd8HDwfSx1xBwC+P8QMGvHAOITxNBSvu/J/mCJvOwv+G4VWkU7zir9SSenTtyCi0LnVtmsc7G5SZo1uV+bxRw==",
       "dev": true,
       "requires": {
-        "acorn": "^6.0.7",
-        "acorn-walk": "^6.1.1",
+        "acorn": "^7.1.1",
```
Comment (ashcrow) [inline client/package-lock.json:13614]: :+1:

File: client/package-lock.json [hunk 9]
```diff
@@ -11391,124 +13557,144 @@
...
         "mkdirp": "^0.5.1",
         "opener": "^1.5.1",
         "ws": "^6.0.0"
       },
       "dependencies": {
         "acorn": {
-          "version": "6.2.1",
-          "resolved": "https://registry.npmjs.org/acorn/-/acorn-6.2.1.tgz",
-          "integrity": "sha512-JD0xT5FCRDNyjDda3Lrg/IxFscp9q4tiYtxE1/nOzlKCk7hIRuYjhq1kCNkbPjMRMZuFq20HNQn1I9k8Oj0E+Q==",
+          "version": "7.1.1",
```
Comment (ashcrow) [inline client/package-lock.json:13630]: :+1:

File: client/package-lock.json [hunk 10]
```diff
@@ -1019,395 +2548,689 @@
...
+        "@intervolga/optimize-cssnano-plugin": "^1.0.5",
+        "@soda/friendly-errors-webpack-plugin": "^1.7.1",
+        "@vue/cli-overlay": "^4.2.3",
+        "@vue/cli-plugin-router": "^4.2.3",
+        "@vue/cli-plugin-vuex": "^4.2.3",
+        "@vue/cli-shared-utils": "^4.2.3",
+        "@vue/component-compiler-utils": "^3.0.2",
+        "@vue/preload-webpack-plugin": "^1.1.0",
+        "@vue/web-component-wrapper": "^1.2.0",
+        "acorn": "^7.1.0",
```
Comment (ashcrow) [inline client/package-lock.json:2830]: :+1:

File: client/package-lock.json [hunk 11]
```diff
@@ -1019,395 +2548,689 @@
...
+            "open": "^6.3.0",
+            "ora": "^3.4.0",
+            "read-pkg": "^5.1.1",
+            "request": "^2.87.0",
+            "request-promise-native": "^1.0.8",
+            "semver": "^6.1.0",
+            "strip-ansi": "^6.0.0"
+          }
+        },
+        "acorn": {
```
Comment (ashcrow) [inline client/package-lock.json:2896]: :+1:

File: client/package-lock.json [hunk 12]
```diff
@@ -7021,9 +9018,9 @@
...
     "json5": {
-      "version": "2.1.0",
-      "resolved": "https://registry.npmjs.org/json5/-/json5-2.1.0.tgz",
-      "integrity": "sha512-8Mh9h6xViijj36g7Dxi+Y4S6hNGV96vcJZr/SrlHh1LR/pEn/8j/+qIBbs44YKl69Lrfctp4QD+AdWLTMqEZAQ==",
+      "version": "2.1.1",
+      "resolved": "https://registry.npmjs.org/json5/-/json5-2.1.1.tgz",
+      "integrity": "sha512-l+3HXD0GEI3huGq1njuqtzYK8OYJyXMkOLtQ53pjWh89tvWS2h6l+1zMkYWqlb57+SiQodKZyvMEFb2X+KrFhQ==",
       "dev": true,
       "requires": {
         "minimist": "^1.2.0"
```
Comment (ashcrow) [inline client/package-lock.json:9026]: :-1:

File: client/package-lock.json [hunk 13]
```diff
@@ -1019,395 +2548,689 @@
...
+        "fs-extra": "^7.0.1",
+        "globby": "^9.2.0",
+        "hash-sum": "^2.0.0",
+        "html-webpack-plugin": "^3.2.0",
+        "launch-editor-middleware": "^2.2.1",
+        "lodash.defaultsdeep": "^4.6.1",
+        "lodash.mapvalues": "^4.6.0",
+        "lodash.transform": "^4.6.0",
+        "mini-css-extract-plugin": "^0.9.0",
+        "minimist": "^1.2.0",
```
Comment (ashcrow) [inline client/package-lock.json:2858]: :-1:  1.2.2+

**Review by ashcrow — CHANGES_REQUESTED** [review]
`minimist` needs updating.

**Comment by zonggen** [comment]
Bumped minimist to 1.2.2, though not sure why the `npm audit` wouldn't show `minimist 1.2.0` as a vulnerability. And `npm install --save minimist@1.2.2` is not updating all of the minimist version. Not sure if we need all of the version to be 1.2.2.

**Comment by zonggen** [comment]
```diff
diff --git a/client/package-lock.json b/client/package-lock.json
index 083f06f..379a885 100644
--- a/client/package-lock.json
+++ b/client/package-lock.json
@@ -9573,9 +9573,9 @@
       }
     },
     "minimist": {
-      "version": "1.2.0",
-      "resolved": "https://registry.npmjs.org/minimist/-/minimist-1.2.0.tgz",
-      "integrity": "sha1-o1AIsg9BOD7sH7kU9M1d95omQoQ=",
+      "version": "1.2.2",
+      "resolved": "https://registry.npmjs.org/minimist/-/minimist-1.2.2.tgz",
+      "integrity": "sha512-rIqbOrKb8GJmx/5bc2M0QchhUouMXSpd1RTclXsB41JdL+VtnojfaJR+h7F9k18/4kHUsBFgk80Uk+q569vjPA==",
       "dev": true
     },
     "minipass": {

```

**Comment by zonggen** [comment]
```
$ npm list minimist

client@0.1.0 /home/abai/git/fcct-online/client
├─┬ @vue/cli-plugin-babel@4.2.3
│ ├─┬ @babel/core@7.8.7
│ │ └─┬ json5@2.1.1
│ │   └── minimist@1.2.2  deduped
│ └─┬ babel-loader@8.0.6
│   └─┬ loader-utils@1.2.3
│     └─┬ json5@1.0.1
│       └── minimist@1.2.2  deduped
├─┬ @vue/cli-service@4.2.3
│ ├── minimist@1.2.2 
│ └─┬ webpack-dev-server@3.10.3
│   └─┬ chokidar@2.1.8
│     └─┬ UNMET OPTIONAL DEPENDENCY fsevents@1.2.9
│       └─┬ UNMET OPTIONAL DEPENDENCY node-pre-gyp@0.12.0
│         ├─┬ UNMET OPTIONAL DEPENDENCY mkdirp@0.5.1
│         │ └── UNMET OPTIONAL DEPENDENCY minimist@0.0.8 
│         └─┬ UNMET OPTIONAL DEPENDENCY rc@1.2.8
│           └── UNMET OPTIONAL DEPENDENCY minimist@1.2.0 
└─┬ eslint@5.16.0
  └─┬ mkdirp@0.5.1
    └── minimist@0.0.8 
```

**Comment by ashcrow** [comment]
All but this one looks good to me:

```
└─┬ eslint@5.16.0
  └─┬ mkdirp@0.5.1
    └── minimist@0.0.8 
```
0.0.8 sounds very old! :fearful:

**Comment by zonggen** [comment]
I don't think I can manually bump version for dependencies used by eslint.. 
Tried to force bump to 1.2.2:
```
└─┬ eslint@5.16.0
  └─┬ mkdirp@0.5.1
    └── minimist@1.2.2  invalid

npm ERR! invalid: minimist@1.2.2 /home/abai/git/fcct-online/client/node_modules/mkdirp/node_modules/minimist
``` 
And the latest eslint version 6.8.0 is still using minimist 0.0.8:
```
└─┬ eslint@6.8.0
  └─┬ mkdirp@0.5.1
    └── minimist@0.0.8 
```

I think it'd be best to check which minimist is creating the security vulnerability and fix that specific dependency version directly.

**Review by ashcrow — APPROVED** [review]

**Comment by ashcrow** [comment]
Do we need `eslint`?

**Comment by zonggen** [comment]
Investigated a little, seems like the `@vue/cli-plugin-babel@4.2.3` use `mkdirp@0.5.1` as part of the dependencies, and removing `eslint` would cause vue not functioning and `npm run serve` failure.

**Comment by ashcrow** [comment]
This gets us further along. I think it's good to go. :+1:

**Comment by zonggen** [comment]
Thank you for reviewing :smiley: 

Merging >>>

**Pull request merged**